File: week2/problem.cpp
class FooBar {
private:
    int n;
    mutex m;
    condition_variable cv;
    bool curFoo = true; 

public:
    FooBar(int n) {
        this->n = n;
    }

    void printorder(bool foo, function<void()>& func) {
        {
            unique_lock lk(m);
            cv.wait(lk, [this, foo](){return curFoo == foo; });
        }
        func();
        {
            unique_lock lk(m);
            curFoo = !curFoo;
        }
        cv.notify_one();
    }

    void foo(function<void()> printFoo) {
        
        for (int i = 0; i < n; i++) {
            printorder(true, printFoo);
        }
    }

    void bar(function<void()> printBar) {
        
        for (int i = 0; i < n; i++) {
            printorder(false, printBar);
        }
    }
};

File: week2/problem.java
class FooBar {
    private int n;
    boolean curFoo = true;

    public FooBar(int n) {
        this.n = n;
    }

    public void printorder(boolean isFoo, Runnable func) throws InterruptedException {
        synchronized (this) {
            while(isFoo != curFoo) {
                wait();
            }
        }
        func.run();
        synchronized (this) {
            curFoo = !curFoo;
            notify();
        }
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        for (int i = 0; i < n; i++) {
           printorder(true, printFoo); 
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            printorder(false, printBar);    
        }
    }
